src/charwise/builder.rs

```diff
@@ -294,136 +293,53 @@ impl CharwiseDoubleArrayAhoCorasickBuilder {
         self.block_len = self.mapper.alphabet_size().next_power_of_two().max(2);
-
         self.states
-            .resize(self.block_len as usize, State::default());
-
-        for i in 0..self.block_len {
-            if i == 0 {
-                self.set_prev(i, self.block_len - 1);
-            } else {
-                self.set_prev(i, i - 1);
-            }
-            if i == self.block_len - 1 {
-                self.set_next(i, 0);
-            } else {
-                self.set_next(i, i + 1);
-            }
-        }
-        self.fix_state(ROOT_STATE_IDX);
-    }
-
-    #[inline(always)]
-    fn fix_state(&mut self, idx: u32) {
-        // DEAD_STATE_IDX-th element should be always free.
-        debug_assert_ne!(idx, DEAD_STATE_IDX);
-        debug_assert!(!self.is_fixed(idx));
-
-        let next = self.get_next(idx);
-        let prev = self.get_prev(idx);
-        self.set_next(prev, next);
-        self.set_prev(next, prev);
-        self.set_fixed(idx);
+            .resize(usize::try_from(self.block_len).unwrap(), State::default());
+        let mut helper = BuildHelper::new(self.block_len, self.num_free_blocks);
+        helper.push_block().unwrap();
+        helper.use_index(ROOT_STATE_IDX);
+        helper.use_index(DEAD_STATE_IDX);
+        helper
     }
 
-    // edges = [(char, next_id)] sorted by char
     #[inline(always)]
-    fn find_base(&self, edges: &[(u32, u32)]) -> u32 {
+    fn find_base(&self, edges: &[(u32, u32)], helper: &BuildHelper) -> u32 {
         debug_assert!(!edges.is_empty());
 
-        let mut idx = self.get_next(DEAD_STATE_IDX);
-        while idx != DEAD_STATE_IDX {
-            debug_assert!(!self.is_fixed(idx));
+        for idx in helper.vacant_iter() {
             let base = idx ^ edges[0].0;
-            if self.verify_base(base, edges) {
+            if self.verify_base(base, edges, helper) {
                 return base;
             }
-            idx = self.get_next(idx);
         }
         u32::try_from(self.states.len()).unwrap() ^ edges[0].0
     }
 
     #[inline(always)]
-    fn verify_base(&self, base: u32, edges: &[(u32, u32)]) -> bool {
+    fn verify_base(&self, base: u32, edges: &[(u32, u32)], helper: &BuildHelper) -> bool {
         for &(c, _) in edges {
             let idx = base ^ c;
-            if self.is_fixed(idx) {
+            if helper.is_used_index(idx) {
                 return false;
             }
         }
         true
     }
 
     #[inline(always)]
-    fn extend_array(&mut self) {
-        let old_len = u32::try_from(self.states.len()).unwrap();
-        let new_len = old_len + self.block_len;
-
-        let num_blocks = old_len / self.block_len;
-        if self.num_free_blocks <= num_blocks {
-            self.close_block(num_blocks - self.num_free_blocks);
+    fn extend_array(&mut self, helper: &mut BuildHelper) -> Result<()> {
+        if u32::try_from(self.states.len()).unwrap() > u32::MAX - self.block_len {
```

Comment:
These large values should be compared using usize to avoid unexpected wrapping.
```suggestion
        if self.states.len() > usize::try_from(u32::MAX - self.block_len).unwrap() {
```

Reply:
Also https://github.com/daac-tools/daachorse/pull/49/commits/38688474206f7e00dd5754c3468f48e5d1d02828